src/peer_loop.rs

```diff
@@ -1821,33 +1823,35 @@ impl PeerLoopHandler {
                 own_datetime_utc.format("%Y-%m-%d %H:%M:%S"));
         }
 
-        // There is potential for a race-condition in the peer_map here, as we've previously
-        // counted the number of entries and checked if instance ID was already connected. But
-        // this check could have been invalidated by other tasks so we perform it again
-
-        if global_state
-            .net
-            .peer_map
-            .values()
-            .any(|pi| pi.instance_id() == self.peer_handshake_data.instance_id)
+        // There is potential for a race-condition in the peer_map here, as
+        // we've previously counted the number of entries and checked if
+        // instance ID was already connected. But other tasks could have made
+        // connections to a new node after the completion of that check, so we
+        // need to make the check while holding a lock to ensure that this
+        // check still holds. Since we're modifyin peer_map if this check is
+        // successful, we have to acquire a read-lock.
```

Comment:
At minimum, this should be “write-lock”, not “read-lock”. However, I think the comment can be improved beyond that.

- There's no race condition “in” the `peer_map`, which is a fully sequential data structure.
- TOCTOU is a well-established class of bug and doesn't have to be explained in full detail; mentioning that the lock is maintained to prevent a TOCTOU bug is enough.

```suggestion
        // Multiple tasks might attempt to set up a connection concurrently. To
        // prevent time-of-check to time-of-use bugs, maintain a write-lock on
        // the global state between the checks and updating the peer map.
```